### PR TITLE
Class to Finetune deep network

### PR DESCRIPTION
**Pull request by stereomatchingkiss**
Provide a class(FineTuneFunction) to finetune the deep network, I write unit test for this class too, the implementation should be correct.

I use MNIST(10000 samples since 60000 take a long time to train), to test the accuracy before finetune and finetune(two sparse autoencoder layers and one softmax layer, the accuracy before finetune is 0.8934, after finetune it raise to 0.9419.

If you think the api are weird, or there are rooms to improve the performance, please give some advices.

### CODE REVIEW

**Comment by stereomatchingkiss** [comment]
Refactor the codes, now the api become

    template<typename EncoderType,
             typename OutputLayerType = regression::SoftmaxRegressionFunction,
             typename DataType = double,
             typename FineTuneGradient = SoftmaxFineTune>
    class FineTuneFunction
    {
     public:        
        /**
         * Construct the class with given data
         * @param input The input data of the LayerTypes and OutputLayerType
         * @param parameters The parameters of the LayerTypes and OutputLayerType
         * @param layerTypes The type(must be tuple) of the Layer(by now only support   SparseAutoencoder)
         * @param outLayerType The type of the last layer(ex : softmax)
         */
        FineTuneFunction(arma::Mat<DataType> &input,
                          EncoderType &encoder,
                          OutputLayerType &outLayerType,
                          arma::Mat<DataType> const &outLayerParam);
    };

construct it like

    auto stackAE = std::forward_as_tuple(sae1, sae2, sae3);
    //cannot get the softmax parameter from SoftmaxRegressionFunction, so we have to pass in
    FineTuneFunction<decltype(stackAE)> finetune(trainData, stackAE, softmaxFunction, softmaxParameter);

However, this class depend on the pull request #451, will wait until #451 is merged

**Comment by stereomatchingkiss** [comment]
I upload the files to my own repository(https://github.com/stereomatchingkiss/mlpack_customize/tree/refactor_autoencoder/src/mlpack/methods/finetune)

Since the other pull request haven't been solved, I cannot commit it ye.The class name FineTuneClass2 is temporary, I use it to compare with the original implementation(FineTuneClass).

**Comment by rcurtin** [comment]
I'm going to close this one for inactivity; I think maybe since it has been almost two years, the ANN API has changed a good bit, so I'm not sure the patch could be applied easily anyway.  @stereomatchingkiss: if you're still interested in working on this, we can reopen it and finish it. :)
